In [1]:
%%capture
%%bash
python ./generate_data.py
python ./run_ddl.py

In [2]:
%%sql --show
use prod.db

26/01/20 04:12:33 WARN SparkSession: Using an existing Spark session; only runtime SQL configurations will take effect.


++
||
++
++

In [3]:
%%sql
show catalogs;

+---------------+
|       catalog |
+---------------+
|          demo |
| spark_catalog |
+---------------+

> In modern catalog systems, the hierarchy is: `Catalog -> Schema -> Table`
<br>A Catalog can have multiple Schemas and each Schema can have multiple tables

In [4]:
%%sql
show schemas IN demo;

+-----------+
| namespace |
+-----------+
|      prod |
+-----------+

In [5]:
%%sql
show schemas IN prod;

-- schema -> namespace

+-----------+
| namespace |
+-----------+
|   prod.db |
+-----------+

In [6]:
%%sql
show tables IN prod.db;

-- namespace -> Table

+-----------+-----------+-------------+
| namespace | tableName | isTemporary |
+-----------+-----------+-------------+
|   prod.db |  customer |       False |
|   prod.db |  lineitem |       False |
|   prod.db |    nation |       False |
|   prod.db |    orders |       False |
|   prod.db |      part |       False |
|   prod.db |  partsupp |       False |
|   prod.db |    region |       False |
|   prod.db |  supplier |       False |
+-----------+-----------+-------------+

In [7]:
%%sql
SELECT
  *
FROM
  prod.db.customer
LIMIT
  2;

+-----------+--------------------+--------------------------------+-------------+-----------------+-----------+--------------+-----------------------------------------------------------------+
| c_custkey |             c_name |                      c_address | c_nationkey |         c_phone | c_acctbal | c_mktsegment |                                                       c_comment |
+-----------+--------------------+--------------------------------+-------------+-----------------+-----------+--------------+-----------------------------------------------------------------+
|         1 | Customer#000000001 |              j5JsirBM9PsCy0O1m |          15 | 25-989-741-2988 |    711.56 |     BUILDING |  y final requests wake slyly quickly special accounts. blithely |
|         2 | Customer#000000002 | 487LW1dovn6Q4dMVymKwwLE9OKf3QG |          13 | 23-768-687-3665 |    121.65 |   AUTOMOBILE | y carefully regular foxes. slyly regular requests about the bli |
+-----------+--------------------+--------------------------------+-------------+-----------------+-----------+--------------+-----------------------------------------------------------------+

In [8]:
%%sql
DESCRIBE lineitem;

-- or prod.db.lineitem if use prod.db not applied during session

+-----------------+---------------+---------+
|        col_name |     data_type | comment |
+-----------------+---------------+---------+
|      l_orderkey |        bigint |    None |
|       l_partkey |        bigint |    None |
|       l_suppkey |        bigint |    None |
|    l_linenumber |           int |    None |
|      l_quantity | decimal(15,2) |    None |
| l_extendedprice | decimal(15,2) |    None |
|      l_discount | decimal(15,2) |    None |
|           l_tax | decimal(15,2) |    None |
|    l_returnflag |        string |    None |
|    l_linestatus |        string |    None |
|      l_shipdate |          date |    None |
|    l_commitdate |          date |    None |
|   l_receiptdate |          date |    None |
|  l_shipinstruct |        string |    None |
|      l_shipmode |        string |    None |
|       l_comment |        string |    None |
+-----------------+---------------+---------+

In [9]:
%%sql
DESCRIBE EXTENDED lineitem;

+------------------------------+------------------------------------------------------------------------------------------------------------------------+---------+
|                     col_name |                                                                                                              data_type | comment |
+------------------------------+------------------------------------------------------------------------------------------------------------------------+---------+
|                   l_orderkey |                                                                                                                 bigint |    None |
|                    l_partkey |                                                                                                                 bigint |    None |
|                    l_suppkey |                                                                                                                 bigint |    None |
|                 l_linenumber |                                                                                                                    int |    None |
|                   l_quantity |                                                                                                          decimal(15,2) |    None |
|              l_extendedprice |                                                                                                          decimal(15,2) |    None |
|                   l_discount |                                                                                                          decimal(15,2) |    None |
|                        l_tax |                                                                                                          decimal(15,2) |    None |
|                 l_returnflag |                                                                                                                 string |    None |
|                 l_linestatus |                                                                                                                 string |    None |
|                   l_shipdate |                                                                                                                   date |    None |
|                 l_commitdate |                                                                                                                   date |    None |
|                l_receiptdate |                                                                                                                   date |    None |
|               l_shipinstruct |                                                                                                                 string |    None |
|                   l_shipmode |                                                                                                                 string |    None |
|                    l_comment |                                                                                                                 string |    None |
|                              |                                                                                                                        |         |
|           # Metadata Columns |                                                                                                                        |         |
|                     _spec_id |                                                                                                                    int |         |
|                   _partition |                                                                                                               struct<> |         |
|                        _file |                                                                                                                 string |         |
|                         _pos |                                

The most common use for querying is to read data from our tables. We can do this using a `SELECT ... FROM` statement<br>

However, running a `SELECT ... FROM` statement can cause issues when the data set is extensive. If you want to examine the data, use `LIMIT n` to instruct **Trino to retrieve only the first n rows**.<br>

We can use the `‘WHERE’ clause` to retrieve rows that match specific criteria. We can specify one or more filters within the`’ WHERE’ clause`. The WHERE clause with more than one filter can use combinations of `AND and OR` criteria to combine the filter criteria, as shown below.

In [10]:
%%sql
-- all customer rows that have c_nationkey = 20 and c_acctbal > 1000
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
  AND c_acctbal > 1000
LIMIT
  10;

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
| c_custkey |             c_name |                                c_address | c_nationkey |         c_phone | c_acctbal | c_mktsegment |                                                                                                          c_comment |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
|         6 | Customer#000000006 |      g1s,pzDenUEBW3O,2 pxu0f9n2g64rJrt5E |          20 | 30-114-968-4951 |   7638.57 |   AUTOMOBILE |            quickly silent asymptotes are slyly regular excuses. instructions wake furiously? quickly bold courts p |
|        81 | Customer#000000081 |                       9jUFbrThIIeoUNd8 9 |          20 | 30-165-277-3269 |   2023.71 |     BUILDING |                                 s against the ironic packages haggle carefully above the slyly express pinto beans |
|       100 | Customer#000000100 |                           MBy6qq3OEGpV4u |          20 | 30-749-445-4907 |   9889.89 |    FURNITURE |                                     dazzle carefully furiously final foxes. express, ironic packages among the qui |
|       210 | Customer#000000210 |                   ,XOlfSzkZDAkm96adR41j, |          20 | 30-876-248-9750 |   7250.14 |    HOUSEHOLD |                                                                              es cajole bravely across the blithely |
|       223 | Customer#000000223 |                    MyQxUcG0P QCetmG00GlF |          20 | 30-193-643-1517 |   7476.20 |     BUILDING |           xcuses. silent theodolites across the carefully bold excuses sleep ironic, final courts. regular excuses |
|       228 | Customer#000000228 | rZ1wxvHNByT71bUJWZjXMDROzlAch6FVu,dj8Zfq |          20 | 30-435-915-1603 |   6868.12 |    FURNITURE |                                                                                      es. blithely permanent sentim |
|       247 | Customer#000000247 |                       eSAW4XaakYFj2WToKU |          20 | 30-151-905-3513 |   8495.92 |    HOUSEHOLD |                               tes nag according to the blithe, even packages. sometimes unusual packages integrate |
|       278 | Customer#000000278 |                         XHAfHlrYQM3elmhJ |          20 | 30-445-570-5841 |   7621.56 |     BUILDING | ely unusual accounts. stealthily special instructions affix blithely. regular, ironic packages sleep even platelet |
|       285 | Customer#000000285 |      rB6fTQKle64k3MvCCatad8DfMgR5OZA G4r |          20 | 30-235-130-1313 |   7276.72 |    FURNITURE |                slyly according to the blithely special instructions. ironic ideas against the blithely furious pac |
|       321 | Customer#000000321 |          LX0SKs3jqo9wH1yixIdGWp2ItclDiuL |          20 | 30-114-675-9153 |   7718.77 |    FURNITURE |                               ng the final, bold requests. furiously regular accounts inside the furiously pending |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+

In [11]:
%%sql
-- All customer rows that have (c_nationkey = 20 and c_acctbal > 1000) or rows that have c_nationkey = 11
SELECT
  *
FROM
  customer
WHERE
  (
    c_nationkey = 20
    AND c_acctbal > 1000
  )
  OR c_nationkey = 11
LIMIT
  10;

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------+
| c_custkey |             c_name |                                c_address | c_nationkey |         c_phone | c_acctbal | c_mktsegment |                                                                                                c_comment |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------+
|         6 | Customer#000000006 |      g1s,pzDenUEBW3O,2 pxu0f9n2g64rJrt5E |          20 | 30-114-968-4951 |   7638.57 |   AUTOMOBILE |  quickly silent asymptotes are slyly regular excuses. instructions wake furiously? quickly bold courts p |
|        52 | Customer#000000052 |              UracAlAA8tSHL5V,poTZIOjh8o, |          11 | 21-186-284-5998 |   5630.28 |    HOUSEHOLD |                                ts boost. carefully express waters across the blithely regular foxes inte |
|        81 | Customer#000000081 |                       9jUFbrThIIeoUNd8 9 |          20 | 30-165-277-3269 |   2023.71 |     BUILDING |                       s against the ironic packages haggle carefully above the slyly express pinto beans |
|        84 | Customer#000000084 |      GB3sUmv RRXV DPzeOSbGxMIF9Z4Eq9 rop |          11 | 21-546-818-3802 |   5174.71 |    FURNITURE |                   ounts. blithely express theodolites nag carefully ironic pinto beans. carefully final  |
|       100 | Customer#000000100 |                           MBy6qq3OEGpV4u |          20 | 30-749-445-4907 |   9889.89 |    FURNITURE |                           dazzle carefully furiously final foxes. express, ironic packages among the qui |
|       131 | Customer#000000131 |            ItdUFrHPZlzjZ, fo03sG4topAKTV |          11 | 21-840-210-3572 |   8595.53 |    HOUSEHOLD |                            ly final Tiresias. slyly permanent theodolites cajole quickly. carefully unus |
|       134 | Customer#000000134 |               6I1TTaoG7bbiogCqRcptG6BYme |          11 | 21-200-159-5932 |   4608.90 |     BUILDING |                                                                    ly regular dolphins haggle blithely.  |
|       148 | Customer#000000148 |                   qJ8bFn4kwiit7RzwGrwo5m |          11 | 21-562-498-6636 |   2135.60 |    HOUSEHOLD |  e carefully pending ideas detect slyly along the furiously special excuses. instructions use carefully  |
|       190 | Customer#000000190 | mY30kK8AfsTGrx,L4zI   QlQnnmCUxikyc8QcZ7 |          11 | 21-730-373-8193 |   1657.46 |   AUTOMOBILE |                                                               y even packages engage furiously pending p |
|       210 | Customer#000000210 |                   ,XOlfSzkZDAkm96adR41j, |          20 | 30-876-248-9750 |   7250.14 |    HOUSEHOLD |                                                                    es cajole bravely across the blithely |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------+

##  SQL Joins Recap

In [12]:
%%sql
SELECT
  COUNT(o.o_orderkey),
  COUNT(l.l_orderkey)
FROM
  orders o
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

+-------------------+-------------------+
| count(o_orderkey) | count(l_orderkey) |
+-------------------+-------------------+
|             24613 |             24613 |
+-------------------+-------------------+

In [13]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  LEFT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|           151933 |               24613 |
+------------------+---------------------+

In [14]:
%%sql
SELECT
  COUNT(o_orderkey)
FROM
  orders;

+-------------------+
| count(o_orderkey) |
+-------------------+
|            150000 |
+-------------------+

In [15]:
%%sql
SELECT
  COUNT(l_orderkey)
FROM
  lineitem;

+-------------------+
| count(l_orderkey) |
+-------------------+
|            600572 |
+-------------------+

The output will include all rows from orders and the rows from lineitem that were able to find at least one matching row from the orders table with the specified join condition (same orderkey and orderdate within a 5-day window of the ship date).

We can also see that the number of rows from the orders table is 151933 & from the lineitem table is 24613. The number of rows in orders is 150000, but the join condition produces 151933 since some orders match with multiple line items.

In [16]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  FULL OUTER JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|           151933 |              600572 |
+------------------+---------------------+

The output will include all rows from orders that match at least one row from the lineitem table with the specified join condition (same orderkey and orderdate within a 5-day window of the ship date) and all rows from the lineitem table.<br>
>Get matched and unmatched rows from both tables

### Cross join - Join every row in the left table with every row in the right table

In [17]:
%%sql
SELECT
  n.n_name AS nation_c_name,
  r.r_name AS region_c_name
FROM
  nation n
  CROSS JOIN region r;

+---------------+---------------+
| nation_c_name | region_c_name |
+---------------+---------------+
|       ALGERIA |        AFRICA |
|       ALGERIA |       AMERICA |
|       ALGERIA |          ASIA |
|       ALGERIA |        EUROPE |
|       ALGERIA |   MIDDLE EAST |
|     ARGENTINA |        AFRICA |
|     ARGENTINA |       AMERICA |
|     ARGENTINA |          ASIA |
|     ARGENTINA |        EUROPE |
|     ARGENTINA |   MIDDLE EAST |
|        BRAZIL |        AFRICA |
|        BRAZIL |       AMERICA |
|        BRAZIL |          ASIA |
|        BRAZIL |        EUROPE |
|        BRAZIL |   MIDDLE EAST |
|        CANADA |        AFRICA |
|        CANADA |       AMERICA |
|        CANADA |          ASIA |
|        CANADA |        EUROPE |
|        CANADA |   MIDDLE EAST |
|         EGYPT |        AFRICA |
|         EGYPT |       AMERICA |
|         EGYPT |          ASIA |
|         EGYPT |        EUROPE |
|         EGYPT |   MIDDLE EAST |
|      ETHIOPIA |        AFRICA |
|      ETHIOPIA |       AMERICA |
|      ETHIOPIA |          ASIA |
|      ETHIOPIA |        EUROPE |
|      ETHIOPIA |   MIDDLE EAST |
|        FRANCE |        AFRICA |
|        FRANCE |       AMERICA |
|        FRANCE |          ASIA |
|        FRANCE |        EUROPE |
|        FRANCE |   MIDDLE EAST |
|       GERMANY |        AFRICA |
|       GERMANY |       AMERICA |
|       GERMANY |          ASIA |
|       GERMANY |        EUROPE |
|       GERMANY |   MIDDLE EAST |
|         INDIA |        AFRICA |
|         INDIA |       AMERICA |
|         INDIA |          ASIA |
|         INDIA |        EUROPE |
|         INDIA |   MIDDLE EAST |
|     INDONESIA |        AFRICA |
|     INDONESIA |       AMERICA |
|     INDONESIA |          ASIA |
|     INDONESIA |        EUROPE |
|     INDONESIA |   MIDDLE EAST |
|          IRAN |        AFRICA |
|          IRAN |       AMERICA |
|          IRAN |          ASIA |
|          IRAN |        EUROPE |
|          IRAN |   MIDDLE EAST |
|          IRAQ |        AFRICA |
|          IRAQ |       AMERICA |
|          IRAQ |          ASIA |
|          IRAQ |        EUROPE |
|          IRAQ |   MIDDLE EAST |
|         JAPAN |        AFRICA |
|         JAPAN |       AMERICA |
|         JAPAN |          ASIA |
|         JAPAN |        EUROPE |
|         JAPAN |   MIDDLE EAST |
|        JORDAN |        AFRICA |
|        JORDAN |       AMERICA |
|        JORDAN |          ASIA |
|        JORDAN |        EUROPE |
|        JORDAN |   MIDDLE EAST |
|         KENYA |        AFRICA |
|         KENYA |       AMERICA |
|         KENYA |          ASIA |
|         KENYA |        EUROPE |
|         KENYA |   MIDDLE EAST |
|       MOROCCO |        AFRICA |
|       MOROCCO |       AMERICA |
|       MOROCCO |          ASIA |
|       MOROCCO |        EUROPE |
|       MOROCCO |   MIDDLE EAST |
|    MOZAMBIQUE |        AFRICA |
|    MOZAMBIQUE |       AMERICA |
|    MOZAMBIQUE |          ASIA |
|    MOZAMBIQUE |        EUROPE |
|    MOZAMBIQUE |   MIDDLE EAST |
|          PERU |        AFRICA |
|          PERU |       AMERICA |
|          PERU |          ASIA |
|          PERU |        EUROPE |
|          PERU |   MIDDLE EAST |
|         CHINA |        AFRICA |
|         CHINA |       AMERICA |
|         CHINA |          ASIA |
|         CHINA |        EUROPE |
|         CHINA |   MIDDLE EAST |
|       ROMANIA |        AFRICA |
|       ROMANIA |       AMERICA |
|       ROMANIA |          ASIA |
|       ROMANIA |        EUROPE |
|       ROMANIA |   MIDDLE EAST |
+---------------+---------------+

In [18]:
%%sql
-- Create a report that shows the number of orders per orderpriority segment.
SELECT
  o_orderpriority,
  COUNT(*)
FROM
  orders
GROUP BY
  o_orderpriority
ORDER BY
  o_orderpriority;

+-----------------+----------+
| o_orderpriority | count(1) |
+-----------------+----------+
|        1-URGENT |    30111 |
|          2-HIGH |    30172 |
|        3-MEDIUM |    29563 |
| 4-NOT SPECIFIED |    29910 |
|           5-LOW |    30244 |
+-----------------+----------+

### Using Subqueries

Create a report that shows the nation, how many items it supplied (by suppliers in that nation), and how many items it purchased (by customers in that nation).

The query below answers the question: how many items it supplied (by suppliers in that nation)

In [19]:
%%sql
SELECT
  n.n_nationkey,
  SUM(l.l_quantity) AS quantity
FROM
  lineitem l
  JOIN supplier s ON l.l_suppkey = s.s_suppkey
  JOIN nation n ON s.s_nationkey = n.n_nationkey
GROUP BY
  n.n_nationkey;

+-------------+-----------+
| n_nationkey |  quantity |
+-------------+-----------+
|          12 | 632809.00 |
|          22 | 719815.00 |
|           1 | 583989.00 |
|          13 | 435841.00 |
|          16 | 523591.00 |
|           6 | 534549.00 |
|           3 | 569306.00 |
|          20 | 720977.00 |
|           5 | 506759.00 |
|          19 | 506500.00 |
|          15 | 624830.00 |
|          17 | 613397.00 |
|           9 | 686980.00 |
|           4 | 610765.00 |
|           8 | 719461.00 |
|          23 | 596263.00 |
|           7 | 765812.00 |
|          10 | 609180.00 |
|          24 | 551367.00 |
|          21 | 596045.00 |
|          11 | 654510.00 |
|          14 | 545734.00 |
|           2 | 664553.00 |
|           0 | 558454.00 |
|          18 | 803315.00 |
+-------------+-----------+

The below query answers the question: how many items it purchased (by customers in that nation)

In [20]:
%%sql
SELECT
  n.n_nationkey,
  SUM(l.l_quantity) AS quantity
FROM
  lineitem l
  JOIN orders o ON l.l_orderkey = o.o_orderkey
  JOIN customer c ON o.o_custkey = c.c_custkey
  JOIN nation n ON c.c_nationkey = n.n_nationkey
GROUP BY
  n.n_nationkey;

+-------------+-----------+
| n_nationkey |  quantity |
+-------------+-----------+
|          12 | 594514.00 |
|          22 | 607446.00 |
|           1 | 609330.00 |
|          13 | 609850.00 |
|          16 | 613443.00 |
|           6 | 585564.00 |
|           3 | 631774.00 |
|          20 | 569819.00 |
|           5 | 647056.00 |
|          19 | 628528.00 |
|          15 | 641135.00 |
|           9 | 659643.00 |
|          17 | 568721.00 |
|           4 | 627841.00 |
|           8 | 584063.00 |
|          23 | 606910.00 |
|           7 | 619356.00 |
|          10 | 678349.00 |
|          24 | 603133.00 |
|          21 | 607776.00 |
|          11 | 625030.00 |
|          14 | 562545.00 |
|           2 | 618707.00 |
|           0 | 600846.00 |
|          18 | 633423.00 |
+-------------+-----------+

In [21]:
%%sql
SELECT
  n.n_name AS nation_c_name,
  s.quantity AS supplied_items_quantity,
  c.quantity AS purchased_items_quantity
FROM
  nation n
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN supplier s ON l.l_suppkey = s.s_suppkey
      JOIN nation n ON s.s_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) s ON n.n_nationkey = s.n_nationkey
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN orders o ON l.l_orderkey = o.o_orderkey
      JOIN customer c ON o.o_custkey = c.c_custkey
      JOIN nation n ON c.c_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) c ON n.n_nationkey = c.n_nationkey;

+----------------+-------------------------+--------------------------+
|  nation_c_name | supplied_items_quantity | purchased_items_quantity |
+----------------+-------------------------+--------------------------+
|          JAPAN |               632809.00 |                594514.00 |
|         RUSSIA |               719815.00 |                607446.00 |
|      ARGENTINA |               583989.00 |                609330.00 |
|         JORDAN |               435841.00 |                609850.00 |
|         FRANCE |               534549.00 |                585564.00 |
|     MOZAMBIQUE |               523591.00 |                613443.00 |
|         CANADA |               569306.00 |                631774.00 |
|   SAUDI ARABIA |               720977.00 |                569819.00 |
|       ETHIOPIA |               506759.00 |                647056.00 |
|        ROMANIA |               506500.00 |                628528.00 |
|        MOROCCO |               624830.00 |                641135.00 |
|      INDONESIA |               686980.00 |                659643.00 |
|           PERU |               613397.00 |                568721.00 |
|          EGYPT |               610765.00 |                627841.00 |
|          INDIA |               719461.00 |                584063.00 |
| UNITED KINGDOM |               596263.00 |                606910.00 |
|        GERMANY |               765812.00 |                619356.00 |
|           IRAN |               609180.00 |                678349.00 |
|  UNITED STATES |               551367.00 |                603133.00 |
|        VIETNAM |               596045.00 |                607776.00 |
|           IRAQ |               654510.00 |                625030.00 |
|          KENYA |               545734.00 |                562545.00 |
|         BRAZIL |               664553.00 |                618707.00 |
|        ALGERIA |               558454.00 |                600846.00 |
|          CHINA |               803315.00 |                633423.00 |
+----------------+-------------------------+--------------------------+

To stack tables on top of each other - use UNION or UNION ALL operator. UNION removes duplicate rows and UNION ALL does not remove duplicate rows.

In [22]:
%%sql
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%_91%'
UNION
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%_91%'
UNION
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%_91%';

+-----------+--------------------+
| c_custkey |             c_name |
+-----------+--------------------+
|      8991 | Customer#000008991 |
|      2916 | Customer#000002916 |
|      7911 | Customer#000007911 |
|      9140 | Customer#000009140 |
|      3291 | Customer#000003291 |
|      5914 | Customer#000005914 |
|      9160 | Customer#000009160 |
|     14991 | Customer#000014991 |
|      9103 | Customer#000009103 |
|     11191 | Customer#000011191 |
|     11991 | Customer#000011991 |
|      9116 | Customer#000009116 |
|      9125 | Customer#000009125 |
|      9186 | Customer#000009186 |
|      9115 | Customer#000009115 |
|      9121 | Customer#000009121 |
|      9158 | Customer#000009158 |
|      9139 | Customer#000009139 |
|      9171 | Customer#000009171 |
|      9197 | Customer#000009197 |
|      9916 | Customer#000009916 |
|     14091 | Customer#000014091 |
|      1910 | Customer#000001910 |
|     10910 | Customer#000010910 |
|     14914 | Customer#000014914 |
|      4915 | Customer#000004915 |
|      7691 | Customer#000007691 |
|      8910 | Customer#000008910 |
|     12915 | Customer#000012915 |
|     13391 | Customer#000013391 |
|      5391 | Customer#000005391 |
|      5891 | Customer#000005891 |
|      5915 | Customer#000005915 |
|      7912 | Customer#000007912 |
|      9100 | Customer#000009100 |
|      9198 | Customer#000009198 |
|     11491 | Customer#000011491 |
|       191 | Customer#000000191 |
|      9162 | Customer#000009162 |
|      2917 | Customer#000002917 |
|      3491 | Customer#000003491 |
|      4391 | Customer#000004391 |
|      3391 | Customer#000003391 |
|      6391 | Customer#000006391 |
|     13791 | Customer#000013791 |
|      4913 | Customer#000004913 |
|      9155 | Customer#000009155 |
|       915 | Customer#000000915 |
|      4918 | Customer#000004918 |
|      2791 | Customer#000002791 |
|      6691 | Customer#000006691 |
|      9194 | Customer#000009194 |
|      9109 | Customer#000009109 |
|      9123 | Customer#000009123 |
|     10917 | Customer#000010917 |
|       691 | Customer#000000691 |
|      9127 | Customer#000009127 |
|      9145 | Customer#000009145 |
|     10911 | Customer#000010911 |
|     11791 | Customer#000011791 |
|       991 | Customer#000000991 |
|     11691 | Customer#000011691 |
|     11910 | Customer#000011910 |
|      3091 | Customer#000003091 |
|      4917 | Customer#000004917 |
|      6491 | Customer#000006491 |
|     13091 | Customer#000013091 |
|      3918 | Customer#000003918 |
|      7791 | Customer#000007791 |
|      9911 | Customer#000009911 |
|     14591 | Customer#000014591 |
|       791 | Customer#000000791 |
|      1691 | Customer#000001691 |
|      9191 | Customer#000009191 |
|      7916 | Customer#000007916 |
|      9129 | Customer#000009129 |
|      9167 | Customer#000009167 |
|      9915 | Customer#000009915 |
|      6791 | Customer#000006791 |
|      8391 | Customer#000008391 |
|      9091 | Customer#000009091 |
|      9148 | Customer#000009148 |
|      9146 | Customer#000009146 |
|      9192 | Customer#000009192 |
|     10991 | Customer#000010991 |
|       491 | Customer#000000491 |
|      5919 | Customer#000005919 |
|     13291 | Customer#000013291 |
|      2491 | Customer#000002491 |
|      4491 | Customer#000004491 |
|      6915 | Customer#000006915 |
|     12912 | Customer#000012912 |
|       916 | Customer#000000916 |
|      3991 | Customer#000003991 |
|      6191 | Customer#000006191 |
|     13591 | Customer#000013591 |
|      7491 | Customer#000007491 |
|      7591 | Customer#000007591 |
|      9914 | Customer#000009914 |
|      5912 | Customer#000005912 |
+-----------+--------------------+

Use `EXCEPT` to subtract tables - When you want to retrieve all the rows from the first table that are not present in the second table

In [23]:
%%sql
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%_91%'
EXCEPT
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%_91%'

+-----------+--------+
| c_custkey | c_name |
+-----------+--------+
+-----------+--------+

The above query has no results as we are subtracting a table with itself so all values from the first table will be present in the second table too hence the query will result in zero records.

In [24]:
%%sql
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%_91%'
EXCEPT
SELECT
  c_custkey,
  c_name
FROM
  customer
WHERE
  c_name LIKE '%191%'

+-----------+--------------------+
| c_custkey |             c_name |
+-----------+--------------------+
|      8991 | Customer#000008991 |
|      2916 | Customer#000002916 |
|      7911 | Customer#000007911 |
|      9140 | Customer#000009140 |
|      3291 | Customer#000003291 |
|      5914 | Customer#000005914 |
|      9160 | Customer#000009160 |
|     14991 | Customer#000014991 |
|      9103 | Customer#000009103 |
|     11991 | Customer#000011991 |
|      9116 | Customer#000009116 |
|      9125 | Customer#000009125 |
|      9186 | Customer#000009186 |
|      9115 | Customer#000009115 |
|      9121 | Customer#000009121 |
|      9158 | Customer#000009158 |
|      9139 | Customer#000009139 |
|      9171 | Customer#000009171 |
|      9197 | Customer#000009197 |
|      9916 | Customer#000009916 |
|     14091 | Customer#000014091 |
|     10910 | Customer#000010910 |
|     14914 | Customer#000014914 |
|      4915 | Customer#000004915 |
|      7691 | Customer#000007691 |
|      8910 | Customer#000008910 |
|     12915 | Customer#000012915 |
|     13391 | Customer#000013391 |
|      5391 | Customer#000005391 |
|      5891 | Customer#000005891 |
|      5915 | Customer#000005915 |
|      7912 | Customer#000007912 |
|      9100 | Customer#000009100 |
|      9198 | Customer#000009198 |
|     11491 | Customer#000011491 |
|      9162 | Customer#000009162 |
|      2917 | Customer#000002917 |
|      3491 | Customer#000003491 |
|      4391 | Customer#000004391 |
|      3391 | Customer#000003391 |
|      6391 | Customer#000006391 |
|     13791 | Customer#000013791 |
|      4913 | Customer#000004913 |
|      9155 | Customer#000009155 |
|       915 | Customer#000000915 |
|      4918 | Customer#000004918 |
|      2791 | Customer#000002791 |
|      6691 | Customer#000006691 |
|      9194 | Customer#000009194 |
|      9109 | Customer#000009109 |
|      9123 | Customer#000009123 |
|     10917 | Customer#000010917 |
|       691 | Customer#000000691 |
|      9127 | Customer#000009127 |
|      9145 | Customer#000009145 |
|     10911 | Customer#000010911 |
|     11791 | Customer#000011791 |
|       991 | Customer#000000991 |
|     11691 | Customer#000011691 |
|      3091 | Customer#000003091 |
|      4917 | Customer#000004917 |
|      6491 | Customer#000006491 |
|     13091 | Customer#000013091 |
|      3918 | Customer#000003918 |
|      7791 | Customer#000007791 |
|      9911 | Customer#000009911 |
|     14591 | Customer#000014591 |
|       791 | Customer#000000791 |
|      1691 | Customer#000001691 |
|      7916 | Customer#000007916 |
|      9129 | Customer#000009129 |
|      9167 | Customer#000009167 |
|      9915 | Customer#000009915 |
|      6791 | Customer#000006791 |
|      8391 | Customer#000008391 |
|      9091 | Customer#000009091 |
|      9148 | Customer#000009148 |
|      9146 | Customer#000009146 |
|      9192 | Customer#000009192 |
|     10991 | Customer#000010991 |
|       491 | Customer#000000491 |
|      5919 | Customer#000005919 |
|     13291 | Customer#000013291 |
|      2491 | Customer#000002491 |
|      4491 | Customer#000004491 |
|      6915 | Customer#000006915 |
|     12912 | Customer#000012912 |
|       916 | Customer#000000916 |
|      3991 | Customer#000003991 |
|     13591 | Customer#000013591 |
|      7491 | Customer#000007491 |
|      7591 | Customer#000007591 |
|      9914 | Customer#000009914 |
|      5912 | Customer#000005912 |
|      9154 | Customer#000009154 |
|     14910 | Customer#000014910 |
|      2911 | Customer#000002911 |
|     11091 | Customer#000011091 |
|      8915 | Customer#000008915 |
|     13913 | Customer#000013913 |
+-----------+--------------------+

In [25]:
%%sql
CREATE VIEW
  nation_supplied_purchased_quantity AS
SELECT
  n.n_name AS nation_c_name,
  s.quantity AS supplied_items_quantity,
  c.quantity AS purchased_items_quantity
FROM
  nation n
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN supplier s ON l.l_suppkey = s.s_suppkey
      JOIN nation n ON s.s_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) s ON n.n_nationkey = s.n_nationkey
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN orders o ON l.l_orderkey = o.o_orderkey
      JOIN customer c ON o.o_custkey = c.c_custkey
      JOIN nation n ON c.c_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) c ON n.n_nationkey = c.n_nationkey;

++
||
++
++

In [26]:
%%sql
SELECT
  *
FROM
  nation_supplied_purchased_quantity;

+----------------+-------------------------+--------------------------+
|  nation_c_name | supplied_items_quantity | purchased_items_quantity |
+----------------+-------------------------+--------------------------+
|          JAPAN |               632809.00 |                594514.00 |
|         RUSSIA |               719815.00 |                607446.00 |
|      ARGENTINA |               583989.00 |                609330.00 |
|         JORDAN |               435841.00 |                609850.00 |
|         FRANCE |               534549.00 |                585564.00 |
|     MOZAMBIQUE |               523591.00 |                613443.00 |
|         CANADA |               569306.00 |                631774.00 |
|   SAUDI ARABIA |               720977.00 |                569819.00 |
|       ETHIOPIA |               506759.00 |                647056.00 |
|        ROMANIA |               506500.00 |                628528.00 |
|        MOROCCO |               624830.00 |                641135.00 |
|      INDONESIA |               686980.00 |                659643.00 |
|           PERU |               613397.00 |                568721.00 |
|          EGYPT |               610765.00 |                627841.00 |
|          INDIA |               719461.00 |                584063.00 |
| UNITED KINGDOM |               596263.00 |                606910.00 |
|        GERMANY |               765812.00 |                619356.00 |
|           IRAN |               609180.00 |                678349.00 |
|  UNITED STATES |               551367.00 |                603133.00 |
|        VIETNAM |               596045.00 |                607776.00 |
|           IRAQ |               654510.00 |                625030.00 |
|          KENYA |               545734.00 |                562545.00 |
|         BRAZIL |               664553.00 |                618707.00 |
|        ALGERIA |               558454.00 |                600846.00 |
|          CHINA |               803315.00 |                633423.00 |
+----------------+-------------------------+--------------------------+

## Exercises
- Write a query that shows the number of items returned for each region name
- List the top 10 most selling parts (part name)
- Sellers (name) who have sold at least one of the top 10 selling parts
- Number of items returned for each order price bucket. The definition of order price bucket is shown below.
```CASE
    WHEN o_totalprice > 100000 THEN 'high'
    WHEN o_totalprice BETWEEN 25000 AND 100000 THEN 'medium'
    ELSE 'low'
END AS order_price_bucket
```

- Average time (in days) between receiptdate and shipdate for each nation (name)

In [27]:
%%sql
show tables IN prod.db;

+-----------+-----------+-------------+
| namespace | tableName | isTemporary |
+-----------+-----------+-------------+
|   prod.db |  customer |       False |
|   prod.db |  lineitem |       False |
|   prod.db |    nation |       False |
|   prod.db |    orders |       False |
|   prod.db |      part |       False |
|   prod.db |  partsupp |       False |
|   prod.db |    region |       False |
|   prod.db |  supplier |       False |
+-----------+-----------+-------------+

Write a query that shows the number of items returned for each region name

In [28]:
%%sql
SELECT
  r.r_name AS region_name,
  COUNT(l.l_returnflag) AS items_returned
FROM
  region r
  JOIN nation n ON r.r_regionkey = n.n_nationkey
  JOIN customer c ON n.n_nationkey = c.c_nationkey
  JOIN orders o ON c.c_custkey = o.o_custkey
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
WHERE
  l.l_returnflag = 'R'
GROUP BY
  r.r_name;

+-------------+----------------+
| region_name | items_returned |
+-------------+----------------+
|        ASIA |           5889 |
|      EUROPE |           6050 |
|     AMERICA |           6045 |
|      AFRICA |           5837 |
| MIDDLE EAST |           6114 |
+-------------+----------------+

List the top 10 most selling parts (part name)

In [29]:
%%sql
SELECT
  p.p_name AS part_name,
  CAST(SUM(l.l_quantity) AS INT) AS total_units_sold
FROM
  part p
  JOIN lineitem l ON p.p_partkey = l.l_partkey
GROUP BY
  1
ORDER BY
  2 DESC
LIMIT
  10

+---------------------------------------+------------------+
|                             part_name | total_units_sold |
+---------------------------------------+------------------+
|        snow blanched blush linen blue |             1484 |
|     moccasin brown puff thistle steel |             1465 |
|   floral azure papaya moccasin indian |             1427 |
| blanched white ghost frosted metallic |             1420 |
|        hot blue honeydew salmon slate |             1413 |
|       lavender green brown linen dark |             1412 |
|   frosted chocolate spring peach lawn |             1398 |
|  hot blanched magenta yellow metallic |             1389 |
|         orange white medium plum drab |             1385 |
|             brown hot olive tan black |             1382 |
+---------------------------------------+------------------+

Sellers (name) who have sold at least one of the top 10 selling parts

In [30]:
%%sql
SELECT DISTINCT
  s.s_name
FROM
  supplier s
  JOIN lineitem l ON s.s_suppkey = l.l_suppkey
WHERE
  l.l_partkey IN (
    SELECT
      l.l_partkey
    FROM
      lineitem l
    GROUP BY
      1
    ORDER BY
      (SUM(l.l_quantity)) DESC
    LIMIT
      10
  )

+--------------------+
|             s_name |
+--------------------+
| Supplier#000000414 |
| Supplier#000000782 |
| Supplier#000000553 |
| Supplier#000000296 |
| Supplier#000000150 |
| Supplier#000000585 |
| Supplier#000000401 |
| Supplier#000000115 |
| Supplier#000000071 |
| Supplier#000000053 |
| Supplier#000000329 |
| Supplier#000000651 |
| Supplier#000000039 |
| Supplier#000000419 |
| Supplier#000000909 |
| Supplier#000000948 |
| Supplier#000000681 |
| Supplier#000000393 |
| Supplier#000000380 |
| Supplier#000000802 |
| Supplier#000000621 |
| Supplier#000000612 |
| Supplier#000000718 |
| Supplier#000000850 |
| Supplier#000000555 |
| Supplier#000000304 |
| Supplier#000000813 |
| Supplier#000000983 |
| Supplier#000000135 |
| Supplier#000000196 |
| Supplier#000000215 |
| Supplier#000000881 |
| Supplier#000000141 |
| Supplier#000000551 |
| Supplier#000000714 |
| Supplier#000000248 |
| Supplier#000000513 |
| Supplier#000000973 |
| Supplier#000000455 |
+--------------------+

Number of items returned for each order price bucket. The definition of order price bucket is shown below.<br>
    ```WHEN o_totalprice > 100000 THEN 'high'
        WHEN o_totalprice BETWEEN 25000 AND 100000 THEN 'medium'
        ELSE 'low'
        END AS order_price_bucket```

In [31]:
%%sql
SELECT ob.order_price_bucket,
        COUNT(*) AS returned_item_count
    FROM
    (
SELECT
    o.o_orderkey,
    CASE
    WHEN o.o_totalprice > 100000 THEN 'high'
    WHEN o.o_totalprice BETWEEN 25000 AND 100000 THEN 'medium'
    ELSE 'low'
END AS order_price_bucket
FROM
 orders o
    ) ob 
JOIN lineitem l
ON ob.o_orderkey = l.l_orderkey
WHERE l.l_returnflag = 'R'
GROUP BY ob.order_price_bucket;

+--------------------+---------------------+
| order_price_bucket | returned_item_count |
+--------------------+---------------------+
|                low |                2745 |
|               high |              121779 |
|             medium |               23777 |
+--------------------+---------------------+

Average time (in days) between receiptdate and shipdate for each nation (name)

In [32]:
%%sql
SELECT
  n.n_name,
  AVG(l.l_receiptdate - l.l_shipdate) AS avg_shipping_delay_days
FROM
  lineitem l
  JOIN supplier s ON l.l_suppkey = s.s_suppkey
  JOIN nation n ON s.s_nationkey = n.n_nationkey
GROUP BY
  n.n_name

+----------------+--------------------------+
|         n_name |  avg_shipping_delay_days |
+----------------+--------------------------+
|          JAPAN | 15 days, 10:13:03.881539 |
|         JORDAN | 15 days, 11:15:12.360606 |
| UNITED KINGDOM | 15 days, 11:37:59.249807 |
|         CANADA | 15 days, 13:34:16.336008 |
|        ALGERIA | 15 days, 11:55:56.133829 |
|        MOROCCO | 15 days, 11:57:32.353305 |
|         FRANCE | 15 days, 12:18:41.196247 |
|     MOZAMBIQUE | 15 days, 11:51:44.272670 |
|          CHINA | 15 days, 12:33:54.926786 |
|   SAUDI ARABIA | 15 days, 11:25:33.652924 |
|      ARGENTINA |  15 days, 9:22:20.118508 |
|           IRAN | 15 days, 12:07:44.594186 |
|         RUSSIA | 15 days, 11:07:44.922132 |
|          INDIA | 15 days, 13:50:39.772404 |
|          KENYA | 15 days, 10:26:27.123863 |
|           IRAQ | 15 days, 13:46:20.483257 |
|        VIETNAM | 15 days, 13:27:37.339233 |
|       ETHIOPIA | 15 days, 12:25:18.503202 |
|      INDONESIA | 15 days, 12:03:28.134454 |
|  UNITED STATES | 15 days, 11:00:19.075704 |
|        ROMANIA | 15 days, 13:14:42.445204 |
|        GERMANY | 15 days, 13:00:10.208507 |
|         BRAZIL | 15 days, 12:12:58.858155 |
|          EGYPT | 15 days, 10:32:11.882372 |
|           PERU | 15 days, 13:26:07.652794 |
+----------------+--------------------------+

In [33]:
%%sql
WITH
  lineitem_delay AS (
    SELECT
      l.l_suppkey,
      l.l_receiptdate - l.l_shipdate AS shipping_delay_days
    FROM
      lineitem l
  ),
  lineitem_with_nation AS (
    SELECT
      ld.shipping_delay_days,
      n.n_name AS nation_name
    FROM
      lineitem_delay ld
      JOIN supplier s ON ld.l_suppkey = s.s_suppkey
      JOIN nation n ON s.s_nationkey = n.n_nationkey
  )
SELECT
  nation_name,
  AVG(shipping_delay_days) AS avg_shipping_delay_days
FROM
  lineitem_with_nation
GROUP BY
  nation_name;

+----------------+--------------------------+
|    nation_name |  avg_shipping_delay_days |
+----------------+--------------------------+
|          JAPAN | 15 days, 10:13:03.881539 |
|         JORDAN | 15 days, 11:15:12.360606 |
| UNITED KINGDOM | 15 days, 11:37:59.249807 |
|         CANADA | 15 days, 13:34:16.336008 |
|        ALGERIA | 15 days, 11:55:56.133829 |
|        MOROCCO | 15 days, 11:57:32.353305 |
|         FRANCE | 15 days, 12:18:41.196247 |
|     MOZAMBIQUE | 15 days, 11:51:44.272670 |
|          CHINA | 15 days, 12:33:54.926786 |
|   SAUDI ARABIA | 15 days, 11:25:33.652924 |
|      ARGENTINA |  15 days, 9:22:20.118508 |
|           IRAN | 15 days, 12:07:44.594186 |
|         RUSSIA | 15 days, 11:07:44.922132 |
|          INDIA | 15 days, 13:50:39.772404 |
|          KENYA | 15 days, 10:26:27.123863 |
|           IRAQ | 15 days, 13:46:20.483257 |
|        VIETNAM | 15 days, 13:27:37.339233 |
|       ETHIOPIA | 15 days, 12:25:18.503202 |
|      INDONESIA | 15 days, 12:03:28.134454 |
|  UNITED STATES | 15 days, 11:00:19.075704 |
|        ROMANIA | 15 days, 13:14:42.445204 |
|        GERMANY | 15 days, 13:00:10.208507 |
|         BRAZIL | 15 days, 12:12:58.858155 |
|          EGYPT | 15 days, 10:32:11.882372 |
|           PERU | 15 days, 13:26:07.652794 |
+----------------+--------------------------+

Calculate the money lost due to discounts. Use the `lineitem` to retrieve the prices of items (excluding discounts) that are part of an order and compare them to the order.

The `l_extendedprice` column does not include discounts & the `o_totalprice` column includes discounts.

In [34]:
%%sql
WITH
  order_item_totals AS (
    SELECT
      l.l_orderkey,
      SUM(l_extendedprice) AS total_item_price
    FROM
      lineitem l
    GROUP BY
      l.l_orderkey
  )
SELECT
  o.o_orderkey,
  o.o_totalprice,
  oit.total_item_price - o.o_totalprice AS discount_loss
FROM
  order_item_totals oit
  JOIN orders o ON oit.l_orderkey = o.o_orderkey

+------------+--------------+---------------+
| o_orderkey | o_totalprice | discount_loss |
+------------+--------------+---------------+
|        964 |    206814.03 |       2439.53 |
|       1697 |    157141.66 |       1660.95 |
|       2214 |    188934.88 |      -3028.52 |
|       2529 |      3889.08 |         19.16 |
|       5409 |    187537.08 |       1252.43 |
|       6721 |    179141.36 |      -2511.39 |
|       7747 |     71926.49 |      -2686.51 |
|       8484 |    110707.18 |       5249.08 |
|      11619 |     42313.46 |       2589.10 |
|      11745 |    154304.82 |       1706.71 |
|      11938 |    153313.46 |       2667.44 |
|      13248 |     46175.82 |        582.79 |
|      13638 |    265504.96 |      -8117.39 |
|      14117 |    327025.28 |       4441.01 |
|      15173 |    225898.83 |       6841.21 |
|      15237 |     52589.27 |      -1511.88 |
|      15846 |    134304.76 |       1659.20 |
|      16742 |     98061.01 |       2477.83 |
|      16896 |    170006.91 |        954.79 |
|      17703 |    161949.49 |       4803.37 |
|      18147 |    127423.98 |       2274.99 |
|      18628 |     46238.51 |      -2992.61 |
|      19141 |     20325.91 |       -767.25 |
|      19907 |    208503.83 |        699.44 |
|      21223 |    212947.77 |       -413.58 |
|      22245 |     88561.33 |       5603.81 |
|      23492 |    274372.24 |      -1270.81 |
|      25479 |     87121.71 |       1464.97 |
|      25826 |    153747.65 |       -999.54 |
|      27651 |    200633.61 |       1463.30 |
|      29089 |    253287.29 |      -3738.70 |
|      29824 |     51496.17 |       -781.82 |
|      30562 |     12701.69 |         28.01 |
|      32098 |     26520.42 |        643.59 |
|      33760 |     57205.82 |       3048.90 |
|      33862 |     49057.17 |       -734.49 |
|      34340 |     31893.80 |       1081.60 |
|      34821 |    289209.01 |      -2403.67 |
|      36102 |    213294.83 |       4474.76 |
|      36103 |    161925.99 |        441.90 |
|      37282 |    167715.61 |       3742.57 |
|      39104 |    122737.87 |        273.29 |
|      39520 |    276702.58 |       -229.98 |
|      39527 |    140212.23 |      -4252.87 |
|      39713 |    197306.45 |       3926.67 |
|      40132 |    383940.24 |        355.22 |
|      41862 |    218957.13 |       3707.97 |
|      41895 |    107458.54 |       2592.94 |
|      41988 |    278537.79 |       7685.83 |
|      42852 |     81639.24 |      -3304.13 |
|      43367 |    125672.87 |       3267.49 |
|      44134 |    177121.39 |      -1511.51 |
|      44901 |     38391.71 |       1394.11 |
|      45410 |    141173.01 |      -1239.73 |
|      47492 |    267452.41 |       4474.26 |
|      47877 |    277658.68 |      10727.29 |
|      48899 |    259147.80 |      -2845.01 |
|      50049 |    183579.42 |       5863.27 |
|      51428 |    316561.63 |       5193.70 |
|      52001 |    370524.22 |      13863.49 |
|      52611 |    171056.50 |       2400.55 |
|      52743 |     85371.94 |      -1278.68 |
|      53509 |    223911.56 |      -3018.81 |
|      54596 |     65009.62 |       2908.88 |
|      54660 |    112723.43 |      -1721.77 |
|      55426 |    138287.82 |       2537.63 |
|      55845 |    174649.81 |      -6256.87 |
|      56964 |    183139.35 |       8018.69 |
|      57157 |     43577.78 |       1967.57 |
|      60033 |     40331.56 |       1702.38 |
|      60738 |    268575.70 |       5348.14 |
|      61251 |     28998.35 |       1208.27 |
|      61888 |    308349.73 |       1812.87 |
|      62273 |    104644.20 |        674.15 |
|      62693 |    214745.19 |      -3490.63 |
|      66114 |    197677.29 |      -2155.02 |
|      68995 |     53001.21 |       3634.81 |
|      69735 |    227910.42 |       -428.76 |
|      70211 |    118535.99 |       4286.80 |
|      70593 |    128255.17 |       2797.28 |
|      71429 |    176950.52 |       7796.47 |
|      71648 |    357682.90 |        963.45 |
|      71655 |    103484.08 |       2313.15 |
|      71936 |    300594.18 |       2883.37 

Sellers (name) who have sold at least one of the top 10 selling parts (use CTE)

In [35]:
%%sql
WITH
  top_10_selling_parts AS (
    SELECT
      l.l_partkey
    FROM
      lineitem l
    GROUP BY
      l.l_partkey
    ORDER BY
      (SUM(l.l_quantity)) DESC
    LIMIT
      10
  )
SELECT DISTINCT
  s.s_name
FROM
  supplier s
  JOIN lineitem l ON s.s_suppkey = l.l_suppkey
  JOIN top_10_selling_parts t ON l.l_partkey = t.l_partkey

+--------------------+
|             s_name |
+--------------------+
| Supplier#000000414 |
| Supplier#000000782 |
| Supplier#000000553 |
| Supplier#000000296 |
| Supplier#000000150 |
| Supplier#000000585 |
| Supplier#000000401 |
| Supplier#000000115 |
| Supplier#000000071 |
| Supplier#000000053 |
| Supplier#000000329 |
| Supplier#000000651 |
| Supplier#000000039 |
| Supplier#000000419 |
| Supplier#000000909 |
| Supplier#000000948 |
| Supplier#000000681 |
| Supplier#000000393 |
| Supplier#000000380 |
| Supplier#000000802 |
| Supplier#000000621 |
| Supplier#000000612 |
| Supplier#000000718 |
| Supplier#000000850 |
| Supplier#000000555 |
| Supplier#000000304 |
| Supplier#000000813 |
| Supplier#000000983 |
| Supplier#000000135 |
| Supplier#000000196 |
| Supplier#000000215 |
| Supplier#000000881 |
| Supplier#000000141 |
| Supplier#000000551 |
| Supplier#000000714 |
| Supplier#000000248 |
| Supplier#000000513 |
| Supplier#000000973 |
| Supplier#000000455 |
+--------------------+

## Window Functions
Window functions allow you to operate on a set of rows at a time and produce output that has the exact grain as the input (vs GROUP BY, which operates on a set of rows, but also changes the meaning of an output row).

Common scenarios when you want to use window functions:

- Calculate running metrics/sliding window over rows in the table (aggregate functions)
- Ranking rows based on values in column(s) (ranking functions)
- Access other rows’ values while operating on the current row (value functions)
- Any combination of the above

In [36]:
%%sql
SELECT
  o_custkey,
  o_orderdate,
  o_totalprice,
  AVG(o_totalprice) -- FUNCTION 
  OVER (
    PARTITION BY
      o_custkey -- PARTITION
    ORDER BY
      o_orderdate -- ORDER BY; ASCENDING ORDER unless specified as DESC
  ) AS running_sum
FROM
  orders
WHERE
  o_custkey = 4
ORDER BY
  o_orderdate
LIMIT
  10;

+-----------+-------------+--------------+---------------+
| o_custkey | o_orderdate | o_totalprice |   running_sum |
+-----------+-------------+--------------+---------------+
|         4 |  1992-06-09 |    261208.46 | 261208.460000 |
|         4 |  1992-07-16 |    206615.30 | 233911.880000 |
|         4 |  1992-10-11 |    222977.75 | 230267.170000 |
|         4 |  1993-08-13 |     80130.69 | 192733.050000 |
|         4 |  1993-11-13 |     66417.98 | 167470.036000 |
|         4 |  1994-03-02 |    274992.65 | 185390.471667 |
|         4 |  1994-04-29 |    199636.40 | 187425.604286 |
|         4 |  1995-05-13 |    107127.51 | 177388.342500 |
|         4 |  1995-08-28 |    282207.37 | 189034.901111 |
|         4 |  1995-09-18 |    181789.20 | 188310.331000 |
+-----------+-------------+--------------+---------------+

Write a query on the orders table that has the following output:

~~~order_month,
o_custkey,
total_price,
three_mo_total_price_avg
consecutive_three_mo_total_price_avg: The consecutive 3 month average of total_price for that customer. Note that this should only include months that are chronologically next to each other.
~~~
>Hint: Use `CAST(strftime(o_orderdate, '%Y-%m-01') AS DATE)` to cast order_month to date format.

>Hint: Use the `INTERVAL` format shown above to construct the window function to compute `consecutive_three_mo_total_price_avg` column.

In [43]:
%%sql
WITH
  monthly_orders AS (
    SELECT
      o.o_custkey,
      CAST(date_format (o.o_orderdate, 'yyyy-M-01') AS DATE) AS order_month,
      SUM(o.o_totalprice) AS total_price
    FROM
      orders o
    GROUP BY
      1,
      2
  )
SELECT
  o_custkey,
  total_price,
  AVG(total_price) OVER (
    PARTITION BY
      o_custkey
    ORDER BY
      order_month ROWS BETWEEN 2 PRECEDING
      AND CURRENT ROW
  ) AS three_mo_total_price_avg
FROM
  monthly_orders
ORDER BY
  1,
  2

+-----------+-------------+--------------------------+
| o_custkey | total_price | three_mo_total_price_avg |
+-----------+-------------+--------------------------+
|         1 |    35523.05 |            221937.106667 |
|         1 |    57669.94 |            124770.333333 |
|         1 |    81485.84 |            117388.036667 |
|         1 |   235155.22 |            177975.270000 |
|         1 |   263247.54 |            315144.135000 |
|         1 |   268835.44 |            135997.073333 |
|         1 |   367040.73 |            367040.730000 |
|         2 |    36433.77 |            109659.980000 |
|         2 |    51599.57 |            123279.743333 |
|         2 |    96852.91 |            146273.085000 |
|         2 |   108967.23 |            118224.476667 |
|         2 |   122500.55 |            214687.806667 |
|         2 |   168931.80 |            109832.866667 |
|         2 |   195693.26 |            195693.260000 |
|         2 |   209272.43 |            114186.370000 |
|         2 |   222236.47 |            208116.660000 |
|         2 |   233181.71 |            151237.693333 |
|         2 |   299326.40 |            251581.526667 |
|         4 |    15671.83 |            175006.933333 |
|         4 |    21012.53 |             36056.000000 |
|         4 |    33563.90 |            126102.803333 |
|         4 |    66417.98 |            123175.473333 |
|         4 |    71483.64 |            118563.443333 |
|         4 |    80130.69 |            169907.913333 |
|         4 |   107127.51 |            193918.853333 |
|         4 |   162955.31 |            208983.960000 |
|         4 |   181789.20 |            190374.693333 |
|         4 |   199636.40 |            180349.010000 |
|         4 |   206615.30 |            233911.880000 |
|         4 |   222977.75 |            230267.170000 |
|         4 |   240814.11 |            282565.953333 |
|         4 |   261208.46 |            261208.460000 |
|         4 |   268534.86 |            263386.193333 |
|         4 |   274992.65 |            140513.773333 |
|         4 |   280809.61 |            213482.550000 |
|         4 |   282207.37 |            196323.760000 |
|         4 |   326074.14 |            174197.783333 |
|         5 |    41715.84 |            108963.080000 |
|         5 |    64473.34 |            136204.073333 |
|         5 |   117887.38 |            173371.986667 |
|         5 |   138421.67 |            166243.150000 |
|         5 |   146751.73 |            159746.010000 |
|         5 |   164780.73 |            145188.523333 |
|         5 |   175977.08 |            189942.270000 |
|         5 |   194064.63 |            194064.630000 |
|         5 |   226251.50 |            189003.103333 |
|         5 |   229069.00 |            139178.856667 |
|         7 |    28009.36 |            151071.803333 |
|         7 |    39757.58 |             66781.163333 |
|         7 |    41899.70 |             81403.523333 |
|         7 |   118686.21 |            111629.140000 |
|         7 |   143727.03 |            143727.030000 |
|         7 |   160837.69 |            197539.786667 |
|         7 |   167413.31 |            163360.246667 |
|         7 |   174301.51 |            121049.520000 |
|         7 |   178940.40 |            161333.715000 |
|         7 |   264368.36 |            203574.023333 |
|         7 |   309189.74 |            241522.570000 |
|         7 |   375620.39 |            178021.393333 |
|         8 |    35543.64 |            198391.330000 |
|         8 |   116649.82 |            116649.820000 |
|         8 |   132600.85 |            178843.536667 |
|         8 |   151811.61 |            158846.346667 |
|         8 |   170212.74 |            198423.076667 |
|         8 |   177090.33 |            206817.570000 |
|         8 |   177287.61 |            236108.790000 |
|         8 |   192126.58 |            234549.720000 |
|         8 |   211803.18 |            229603.396667 |
|         8 |   299719.40 |            279277.676667 |
|         8 |   308406.67 |            2

### Ranking functions

Get the top 3 spending customers per day

In [44]:
%%sql
WITH
  ranking AS (
    SELECT
      o_orderdate,
      o_totalprice,
      o_custkey,
      RANK() OVER (
        PARTITION BY
          o_orderdate
        ORDER BY
          o_totalprice DESC
      ) AS rnk
    FROM
      orders
  )
SELECT
  *
FROM
  ranking
WHERE
  rnk <= 3
ORDER BY
  o_orderdate
LIMIT
  5;

+-------------+--------------+-----------+-----+
| o_orderdate | o_totalprice | o_custkey | rnk |
+-------------+--------------+-----------+-----+
|  1992-01-01 |    358534.43 |     10942 |   1 |
|  1992-01-01 |    307146.32 |      5426 |   2 |
|  1992-01-01 |    306169.25 |     12973 |   3 |
|  1992-01-02 |    319926.05 |     12161 |   1 |
|  1992-01-02 |    302943.34 |     10526 |   2 |
+-------------+--------------+-----------+-----+

In [45]:
%%sql
-- Let's look at an example showing the difference between RANK, DENSE_RANK and ROW_NUMBER
SELECT
  order_date,
  order_id,
  total_price,
  ROW_NUMBER() OVER (
    PARTITION BY
      order_date
    ORDER BY
      total_price
  ) AS ROW_NUMBER,
  RANK() OVER (
    PARTITION BY
      order_date
    ORDER BY
      total_price
  ) AS RANK,
  DENSE_RANK() OVER (
    PARTITION BY
      order_date
    ORDER BY
      total_price
  ) AS DENSE_RANK
FROM
  (
    SELECT
      '2024-07-08' AS order_date,
      'order_1' AS order_id,
      100 AS total_price
    UNION ALL
    SELECT
      '2024-07-08',
      'order_2',
      200
    UNION ALL
    SELECT
      '2024-07-08',
      'order_3',
      150
    UNION ALL
    SELECT
      '2024-07-08',
      'order_4',
      90
    UNION ALL
    SELECT
      '2024-07-08',
      'order_5',
      100
    UNION ALL
    SELECT
      '2024-07-08',
      'order_6',
      90
    UNION ALL
    SELECT
      '2024-07-08',
      'order_7',
      100
    UNION ALL
    SELECT
      '2024-07-10',
      'order_8',
      100
    UNION ALL
    SELECT
      '2024-07-10',
      'order_9',
      100
    UNION ALL
    SELECT
      '2024-07-10',
      'order_10',
      100
    UNION ALL
    SELECT
      '2024-07-11',
      'order_11',
      100
  ) AS orders
ORDER BY
  order_date,
  ROW_NUMBER;

+------------+----------+-------------+------------+------+------------+
| order_date | order_id | total_price | ROW_NUMBER | RANK | DENSE_RANK |
+------------+----------+-------------+------------+------+------------+
| 2024-07-08 |  order_4 |          90 |          1 |    1 |          1 |
| 2024-07-08 |  order_6 |          90 |          2 |    1 |          1 |
| 2024-07-08 |  order_1 |         100 |          3 |    3 |          2 |
| 2024-07-08 |  order_5 |         100 |          4 |    3 |          2 |
| 2024-07-08 |  order_7 |         100 |          5 |    3 |          2 |
| 2024-07-08 |  order_3 |         150 |          6 |    6 |          3 |
| 2024-07-08 |  order_2 |         200 |          7 |    7 |          4 |
| 2024-07-10 |  order_8 |         100 |          1 |    1 |          1 |
| 2024-07-10 |  order_9 |         100 |          2 |    1 |          1 |
| 2024-07-10 | order_10 |         100 |          3 |    1 |          1 |
| 2024-07-11 | order_11 |         100 |          1 |    1 |          1 |
+------------+----------+-------------+------------+------+------------+

Write a query on the orders table that has the following output:
- o_custkey
- order_month: In YYYY-MM format, use strftime(o_orderdate, ‘%Y-%m’) AS order_month
- total_price: Sum of o_totalprice for that month
- three_mo_total_price_avg: The 3 month (previous, current & next) average of total_price for that customer

In [51]:
%%sql
SELECT
  o_custkey,
  order_month,
  total_price,
  AVG(total_price) OVER (
    PARTITION BY
      o_custkey
    ORDER BY
      order_month ROWS BETWEEN 1 PRECEDING
      AND 1 FOLLOWING
  ) AS three_mo_total_price_avg
FROM
  (
    SELECT
      o_custkey,
      date_format (o_orderdate, 'yyyy-MM') AS order_month,
      SUM(o_totalprice) AS total_price
    FROM
      orders
    GROUP BY
      1,
      2
  )

+-----------+-------------+-------------+--------------------------+
| o_custkey | order_month | total_price | three_mo_total_price_avg |
+-----------+-------------+-------------+--------------------------+
|         1 |     1992-04 |   367040.73 |            315144.135000 |
|         1 |     1993-06 |   263247.54 |            221937.106667 |
|         1 |     1994-12 |    35523.05 |            177975.270000 |
|         1 |     1995-11 |   235155.22 |            117388.036667 |
|         1 |     1996-05 |    81485.84 |            124770.333333 |
|         1 |     1996-12 |    57669.94 |            135997.073333 |
|         1 |     1997-03 |   268835.44 |            163252.690000 |
|         5 |     1992-04 |   194064.63 |            166243.150000 |
|         5 |     1992-05 |   138421.67 |            159746.010000 |
|         5 |     1994-06 |   146751.73 |            108963.080000 |
|         5 |     1995-03 |    41715.84 |            139178.856667 |
|         5 |     1995-07 |   229069.00 |            145188.523333 |
|         5 |     1995-10 |   164780.73 |            189942.270000 |
|         5 |     1997-01 |   175977.08 |            189003.103333 |
|         5 |     1997-03 |   226251.50 |            173371.986667 |
|         5 |     1997-10 |   117887.38 |            136204.073333 |
|         5 |     1998-06 |    64473.34 |             91180.360000 |
|         7 |     1992-05 |   143727.03 |            161333.715000 |
|         7 |     1992-08 |   178940.40 |            163360.246667 |
|         7 |     1993-06 |   167413.31 |            203574.023333 |
|         7 |     1994-02 |   264368.36 |            197539.786667 |
|         7 |     1994-03 |   160837.69 |            151071.803333 |
|         7 |     1994-10 |    28009.36 |            121049.520000 |
|         7 |     1995-03 |   174301.51 |             81403.523333 |
|         7 |     1995-07 |    41899.70 |            111629.140000 |
|         7 |     1995-09 |   118686.21 |             66781.163333 |
|         7 |     1996-05 |    39757.58 |            178021.393333 |
|         7 |     1996-11 |   375620.39 |            241522.570000 |
|         7 |     1998-04 |   309189.74 |            342405.065000 |
|        10 |     1992-04 |    49364.56 |             87422.340000 |
|        10 |     1992-07 |   125480.12 |            126153.033333 |
|        10 |     1993-07 |   203614.42 |            213859.226667 |
|        10 |     1993-09 |   312483.14 |            245072.583333 |
|        10 |     1994-01 |   219120.19 |            201163.413333 |
|        10 |     1994-05 |    71886.91 |            172559.306667 |
|        10 |     1994-10 |   226670.82 |            154314.463333 |
|        10 |     1995-06 |   164385.66 |            191886.423333 |
|        10 |     1996-04 |   184602.79 |            179618.473333 |
|        10 |     1996-05 |   189866.97 |            253722.580000 |
|        10 |     1997-09 |   386697.98 |            273291.170000 |
|        10 |     1998-01 |   243308.56 |            239919.346667 |
|        10 |     1998-03 |    89751.50 |            166530.030000 |
|        17 |     1993-03 |   100506.44 |            151178.455000 |
|        17 |     1993-05 |   201850.47 |            133334.373333 |
|        17 |     1993-06 |    97646.21 |            129004.763333 |
|        17 |     1993-09 |    87517.61 |             82454.146667 |
|        17 |     1995-02 |    62198.62 |            124389.740000 |
|        17 |     1997-12 |   223452.99 |            142825.805000 |
|        19 |     1992-08 |    24793.63 |            102551.360000 |
|        19 |     1992-11 |   180309.09 |            130880.623333 |
|        19 |     1993-02 |   187539.15 |            198045.496667 |
|        19 |     1994-05 |   226288.25 |            164017.736667 |
|        19 |     1995-06 |    78225.81 |            204083.690000 |
|        19 |     1995-09 |   307737.01 |            220413.490000 |
|        19 |     1996-02 |   275277.65 |            205694.606667 

From the orders table get the 3 lowest spending customers per day

In [55]:
%%sql
WITH
  ranking AS (
    SELECT
      o_custkey,
      o_totalprice,
      o_orderdate,
      RANK() OVER (
        PARTITION BY
          o_orderdate
        ORDER BY
          o_totalprice
      ) AS rnk
    FROM
      orders
  )
SELECT
  *
FROM
  ranking
WHERE
  rnk <= 3
LIMIT
  6;

+-----------+--------------+-------------+-----+
| o_custkey | o_totalprice | o_orderdate | rnk |
+-----------+--------------+-------------+-----+
|      9376 |     12781.38 |  1992-01-01 |   1 |
|      9103 |     25496.77 |  1992-01-01 |   2 |
|     14218 |     30944.67 |  1992-01-01 |   3 |
|      2746 |     16908.46 |  1992-01-02 |   1 |
|      4355 |     25445.96 |  1992-01-02 |   2 |
|      5311 |     26063.75 |  1992-01-02 |   3 |
+-----------+--------------+-------------+-----+

Write a SQL query using the orders table that calculates the following columns:
- o_orderdate: From orders table
- o_custkey: From orders table
- o_totalprice: From orders table
- totalprice_diff: The customers current day’s o_totalprice - that same customers most recent previous purchase’s o_totalprice

>Hint: Start by figuring out what the PARTITION BY column should be, then what the ORDER BY column should be, and then finally the function to use.

>Hint: Use the LAG(column_name) ranking function to identify the prior day’s revenue.